src/ORM/Search/SearchContextForm.php

```diff
@@ -0,0 +1,210 @@
+            }, array_keys($searchParams ?? [])), $searchParams ?? []);
+        }
+
+        $schema = [
+            'formSchemaUrl' => $schemaUrl,
+            'name' => $this->getSearchField(),
+            // stdClass will map to empty json object '{}' when encoded
+            'filters' => $searchParams ?: new \stdClass()
+        ];
+
+        // GridField has some special requirements to ensure the "state" of the filter header is updated
+        $parentRequestHandler = $this->getController();
+        if ($parentRequestHandler instanceof GridField) {
+            $searchAction = GridField_FormAction::create($parentRequestHandler, 'filter', false, 'filter', null);
+            $clearAction = GridField_FormAction::create($parentRequestHandler, 'reset', false, 'reset', null);
+            $schema = array_merge($schema, [
+                'gridfield' => $parentRequestHandler->getName(),
+                'searchAction' => $searchAction->getAttribute('name'),
+                'clearAction' => $clearAction->getAttribute('name'),
+            ]);
+            $filterHeader = $parentRequestHandler->getConfig()->getComponentByType(GridFieldFilterHeader::class);
+            if ($filterHeader) {
+                $placeholder = $filterHeader->getPlaceHolderText();
+                if ($placeholder) {
+                    $schema['placeholder'] = $placeholder;
+                }
+            }
+        }
+
+        if (!isset($schema['placeholder'])) {
+            $schema['placeholder'] = _t(
+                __CLASS__ . '.FILTERLABELTEXT',
+                'Search "{model}"',
+                ['model' => ClassInfo::hasMethod($singleton, 'i18n_plural_name') ? $singleton->i18n_plural_name() : ClassInfo::shortName($singleton)]
+            );
+        }
+
+        return $schema;
+    }
+
+    /**
+     * Get the rendered placeholder with schema data required for generating the search filter form in the CMS
+     */
+    public function getPlaceHolder(bool $isFiltered): DBHTMLText
+    {
+        $templateCandidates = SSViewer::get_templates_by_class(static::class, '_Placeholder', __CLASS__);
+        return $this->renderWith($templateCandidates, ['IsFiltered' => $isFiltered]);
+    }
+
+    /**
+     * Get the rendered search filter button which toggles display of the search filter form in the CMS
+     */
+    public function getFilterButton(bool $isFiltered): DBHTMLText
+    {
+        $templateCandidates = SSViewer::get_templates_by_class(static::class, '_Button', __CLASS__);
+        return $this->renderWith($templateCandidates, ['IsFiltered' => $isFiltered]);
+    }
+
+    /**
+     * Takes raw values submitted e.g. via AJAX and pulls them through FormField instances.
+     * This allows form fields such as CurrencyField to make any relevant adjustments prior to using
+     * the values in the SearchContext database comparisons.
+     */
+    public function prepareValuesForSearchContext(array $values): array
```

Comment:
Does this need to be public? Seems like it should be private

Reply:
Actually nevermind, I see it's used [here](https://github.com/silverstripe/silverstripe-cms/pull/3052/files#diff-8de4622cc87cca016e9e2d317482ab37ac8a29a900134ad8ffad3d89ee108937R1547)